RFC-0004-jdbc-join-pushdown.md

```diff
@@ -0,0 +1,429 @@
+
+If we  "Push down" or send these joins which involves on same catalog/remote datasource as part of the  SQL along with join condition to the remote data source it increase the performance 3x to 10x.
+
+
+
+## Background
+
+
+
+This implementation is to address a performance limitation of Presto federation of SQLs of JDBC connector to remote data sources such as DB2, Postgres, Oracle etc. Currently, in Presto, we have predicate pushdown (WHERE condition pushdown) to some extent in JDBC connectors and not having any join pushdown or join condition pushdown capabilities. 
+
+This cause high performance impact on join queries and it is raised by some of our client. While comparing with competitors we also missing the Jdbc join pushdown capabilities.
+
+We did a poc by changing the presto generated PlanNode to handle jdbc join pushdown and it increases the performance from 3x on postgres and  8x on db2 remote datasource. Now we need to perform its actual implementation 
+
+
+
+## Proposed Implementation
+
+If presto get a join query which is trying to join tables either from same datasource or from different datasource, it is receiving as a string formatted sql query. Using presto parser and analyser, presto validate the syntax and converted to Query (Statement) object. This Query object is converted to presto internal reference architecture called Plan, using its logical and physical optimizers. Finally this plan is executed by the executor. 
+
+Currently for the join query, presto create a final plan which contains separate TableScanNode for each table that participated on join query and this TableScanNode info is used by the connector to create the select query. On top of this select query result, presto apply join condition and other predicate to provide the final result.
+
+In the proposed implementation, while performing the logical optimization, instead of creating separate TableScanNode for each table of the JoinNode, it uses a new single TableScanNode for holding all the table details which satisfies the "JoinPushdown condition" (JoinPushdown condition pointed below) in the join query. This new TableScanNode also hold join criteria for that "JoinPushdown condition" satisfied tables. Using this new TableScanNode it build join query at connector level and return the join result to presto. Now the further predicate which was not pushed down to connector level will apply on the result by the presto and return the final result.
+
+#### The JoinPushdown condition
+
+ 1. Table: left table and right table should be on same connector
+ 2. Join clause: Join condition should be from same connectors table
+ 3. Pushdown flag: A global setting is there for enable JdbcJoinPushdown
+ 4. Filter Criteria: The filter criteria should be from same connector
```

Comment:
What are these ? Would be great to illustrate this with an example.

Reply:
This is updated with an example query